### PR TITLE
starting new data structures

### PR DESCRIPTION
**Pull request by aepaysinger**

### CODE REVIEW

**Comment by aepaysinger** [comment]
for this part: 
len(the_list) returns the size of the list
print(the_list) returns what the display() method returns
do i need to adjust __repr__ and __len__ in the node function?

File: linkedlist.py [hunk 1]
```diff
@@ -19,29 +18,69 @@ def pop(self):
...
             
+        
 
 
 class Node():
-    def __init__(self, value, next):
+    def __init__(self, value, next, previous):
         self.value = value
         self.next = next
+        self.previous = previous
```
Comment (W-Ely) [inline linkedlist.py]: Could link to the Data Structure you are doing?  If this is a Regular linked list, there is no `previous`. The linking is one directional. If you are working on the doubly linked list the items are linked both directions.

**Comment by W-Ely** [comment]
> for this part:
len(the_list) returns the size of the list
print(the_list) returns what the display() method returns
do i need to adjust repr and len in the node function?

I don't see a function called `node`, could you restate the question?

**Comment by aepaysinger** [comment]
so i put the __repr__ method in the class LinkedList. and if i create a linked list and the print it i get what I want but I am having issues with the test for it passing

File: test_linkedlist.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    with pytest.raises(ValueError) as e_info:
+        ll.remove(ll.head.next.next.next)
+    assert e_info.value.args[0] == "node not in LinkedList"
+
+
+def test_print():
+    ll = LinkedList([5, 27, 8])
+
+    assert print(ll) == "(5, 27, 8)"
```
Comment (W-Ely) [inline test_linkedlist.py]: Print is a function, what does it return?

```python
a = print("123")
print(a) #  <--- what print _does_ and what it _returns_ are 2 different things.
``` 

How can you test what it does, not what it returns?  What is it actually doing?  What does it mean to print?  Hint [stdout](https://stackoverflow.com/questions/33767627/python-write-unittest-for-console-print).

File: linkedlist.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            self.head = self.head.next
+            self._length -= 1
+        else:
+            raise ValueError("Empty List")
+        return old_head.value
+
+    def size(self):
+        return self._length
+
+    def search(self, val):
```
Comment (W-Ely) [inline linkedlist.py:27]: If I have `val` and can call `ll.search(val)` why would I want it to return the `val`?  check the instructions.

**Comment by W-Ely** [comment]
Also update the readme!

File: README.md [hunk 4]
```diff
@@ -1 +1,8 @@
-# data-structures-new
\ No newline at end of file
+# data-structures-new
+## LinkedList:
+    * push(val): Inserts the value at the head of the list
+    * pop(): Pops the first value off the head of the list and returns it. Raises an exception if there are no values to return.
+    * size(): Returns the length of the list.
+    * search(val): Returns the node containing the val in the list if it is present. If it is not present it will return None.
+    * remove(node): Removes the given node from the list. If the node is not in the list is will raise an exception.
+    * display(): Returns a unicode string representing the list as if it were a Pthon tuple literal. Ex: "(5, 'Bravo', 9)"
```
Comment (W-Ely) [inline README.md]: len(the_list) returns the size of the list
print(the_list) returns what the display() method returns

File: linkedlist.py (hunk 3 above)
Comment (aepaysinger) [reply linkedlist.py:27]: you are right. it should return the node not value

**Review by W-Ely — APPROVED** [review]

**Pull request merged**